unstable/wlr-exec-secure-unstable-v1.xml

```diff
@@ -0,0 +1,139 @@
+        choose to disallow this request for certain commands, in which case a
+        protocol error is issued. The command will be executed without a shell.
+      </description>
+      <arg name="handle" type="new_id" interface="zwlr_exec_handle_v1"/>
+      <arg name="command" type="array" summary="the command to execute"/>
+    </request>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the manager">
+        All objects created by the manager will still remain valid, until their
+        appropriate destroy request has been called.
+      </description>
+    </request>
+  </interface>
+
+  <interface name="zwlr_exec_handle_v1" version="1">
+    <description summary="an executing or to-be-executed process">
+      This resource represents a process executed by the compositor, or a
+      process which will be executed by the compositor shortly.
+
+      The client should transfer any appropriate state, then send an exec
+      request and await the exit event.
+
+      Once the compositor receives an exec request, it should fork, close all
+      file descriptors (except for the transferred ones), then dup2 them to the
+      appropriate file descriptor number. It should then exec (with PATH lookup,
+      e.g. via execlp or exevp) the requested command. Then it should send the
+      pid event, followed by the exit event when the child terminates.
+    </description>
+
+    <request name="environment">
+      <description summary="set an environment variable in the process">
+        Sets an environment variable for the new process. The compositor can
+        choose not to set any variables, including for example LD_PRELOAD.
+      </description>
+      <arg name="name" type="string" summary="environment variable name"/>
+      <arg name="value" type="string" summary="environment variable value"/>
+    </request>
+
+    <request name="fd">
+      <description summary="transfer a file descriptor to the process">
+        Transfers a file descriptor to the new process. The compositor will use
+        dup2 to assign the specified file descriptor number, then close the
+        danging fd.
+      </description>
+      <arg name="fd" type="fd" summary="file descriptor to transfer"/>
+      <arg name="num" type="uint" summary="first parameter to dup2"/>
+    </request>
+
+    <request name="exec">
+      <description summary="start the new process">
+        Asks the compositor to start the new process. Any request other than
+        destroy after this point is a protocol violation.
+      </description>
+    </request>
+
+    <event name="pid">
+      <description summary="the process's pid is available">
+        The compositor sends this after it has forked, to inform the client of
+        the new process's pid.
+      </description>
+      <arg name="pid" type="uint"
+           summary="pid_t of the child process"/>
+    </event>
+
+    <event name="exit">
```

Comment:
I don't like the idea of this event. It basically asks the compositor to become a service manager, and track extra client state over normal wayland clients. It just seems unnecessary in my opinion.

It also doesn't address harder questions about processes that fork themselves (daemon-like) or otherwise hand their WAYLAND_SOCKET to another process. The compositor doesn't have a way to track this kind of state and shouldn't be expected to.

Also, there is an indefinite amount of time before this is called. How long do we really expect launcher programs to wait around doing nothing? It would just be cleaner if you fire off the client and don't hear about it again.

As far as I can see, there are two potential use cases for this extension:

### (General) Application launchers

As stated above, there isn't really a robust way to track the state of the clients (unless you want to implement/depend on full-blown service manager), and application launchers wouldn't care anyway. They would ignore this event, and destroy the object.

### Applications gaining privileges for themselves

Perhaps a program wants to provide a proper interactive interface and really do care about the state of the process they create. We don't need the exit event to be able to implement this.
The fd/enviroment arguments allows clients to implement anything special it wants.

You can create a socketpair/pipe in the launcher process, pass it to the child process, and communicate anything you want back via that, including noticing when the socket/pipe closes (e.g. on death).

Reply:
>It also doesn't address harder questions about processes that fork themselves (daemon-like) or otherwise hand their WAYLAND_SOCKET to another process. The compositor doesn't have a way to track this kind of state and shouldn't be expected to.

It's not becoming a service manager because I don't expect it to do this, either. This is basically fork/exec/wait the protocol, nothing more and it doesn't need to handle the edge cases. Clients using this and launched by this will be designed to accomodate it, the edge cases are not important.

>Perhaps a program wants to provide a proper interactive interface and really do care about the state of the process they create. We don't need the exit event to be able to implement this.
>
>The fd/enviroment arguments allows clients to implement anything special it wants.

I disagree, pid is a natural thing to include in this design and its absence would be glaring. Making applications deal with some secondary exit hatch which sends an exit status down the drain first is dumb. The compositor will already have an event loop and handling SIGCHLD isn't a tall order.